### ISSUE TITLE
GeneralMixtureModel.probability() returns with impossible answers

### ISSUE THREAD

**Issue by DeanLa** [jmschrei/pomegranate#498]
I have a GMM (full json in the end).
When I create a histogram of 100000 it looks like this:
![image](https://user-images.githubusercontent.com/7852981/46745437-b2759f80-ccb5-11e8-8a21-55a6c02283a2.png)

but I try to plot the same with .probability I get impossible answers
```python
In[249]: z.probability([1, 1.5])
Out[254]: array([ 63493.63593432,      0.        ])
```
It just returns with 0 almost everywhere and a very large number for 1.
the plot looks like this:
![image](https://user-images.githubusercontent.com/7852981/46745365-9540d100-ccb5-11e8-9e51-796d4e259930.png)

When I ran the examples from the notebooks (With other GMM) it worked as expected.
I run on version 0.9.0

Full GMM
```json
{
    "class" : "GeneralMixtureModel",
    "distributions" : [
        {
            "class" : "Distribution",
            "name" : "LogNormalDistribution",
            "parameters" : [
                3.3485025325681843e-124,
                0.01
            ],
            "frozen" : false
        },
        {
            "class" : "Distribution",
            "name" : "PoissonDistribution",
            "parameters" : [
                2.244755806372583
            ],
            "frozen" : false
        },
        {
            "class" : "Distribution",
            "name" : "NormalDistribution",
            "parameters" : [
                1.7412781885356912,
                0.383781154574139
            ],
            "frozen" : false
        }
    ],
    "weights" : [
        -0.629690822505316,
        -4.6881319295791535,
        -0.7807995384304394
    ]
}
```

**Comment by jmschrei** [comment]
Howdy

I think you're running into an issue with the Poisson distribution that yields 0 probabilities for non integer numbers. When I try to run with the most recent version I get the following:

```python
>>> model
{
    "weights" : [
        0.103251323415235,
        0.7687198364244546,
        0.1280288401603104
    ],
    "distributions" : [
        {
            "frozen" : false,
            "class" : "Distribution",
            "parameters" : [
                3.34850253257e-124,
                0.01
            ],
            "name" : "LogNormalDistribution"
        },
        {
            "frozen" : false,
            "class" : "Distribution",
            "parameters" : [
                2.24475580637
            ],
            "name" : "PoissonDistribution"
        },
        {
            "frozen" : false,
            "class" : "Distribution",
            "parameters" : [
                1.74127818854,
                0.383781154574
            ],
            "name" : "NormalDistribution"
        }
    ],
    "class" : "GeneralMixtureModel"
}
>>> model.probability([1, 1.5])
array([4.32257077, 0.31528453])
```

These numbers seem more reasonable. Remember that probabilities can be above 1. Consider a uniform distribution between 0 and 2. Each point between 0 and 2 will have a probability of 0.5. Likewise, a uniform distribution between 0 and 0.5 will yield a probability of 2 at each point between 0 and 0.5. This same thing can happen with normal or log normal distribution with a very small variance.

Can you try updating and see whether that fixes your bug?

Jacob

**Comment by jmschrei** [comment]
Actually it looks like I might've changed the way the weights are stored since the version you're using. You're storing the `log(weights)`, which are negative, whereas now they're always positive probabilities. Even when correcting for that (replacing `weights` with `numpy.exp(weights) / numpy.exp(weights).sum()`) I still get a reasonable answer: `[21.32982996  0.39322031]`

**Comment by jmschrei** [comment]
Can you expand a little bit on what you man by "tried re-running with the normalized exponent"? What I mean is that the weight attribute in 0.9.0 stores the log weight rather than the actual weight, and I think this was corrected in 0.10.0.

If you can't upgrade I'd suggest not using the Poisson distribution because it appears to be broken. Perhaps ExponentialDistribution or LogNormalDistribution would be sufficient?

**Comment by DeanLa** [comment]
The normalized exponent was `numpy.exp(weights) / numpy.exp(weights).sum()` where `weights` is the log weight. The new weights were ~ `array([ 0.53, 0.0092, 0.46])` where in your example the weights were `np.array(0.10, 0.76, 0.12)` which is `-weights/-weights.sum()` so I wondered why is that.
